core/tee/tee_fs_common.c

```diff
@@ -250,63 +1236,141 @@ int tee_fs_common_write(struct tee_fs_fd *fdp,
+ *    created folder (new TEE file)
+ *  - Now we are ready to commit meta, create hard link for the
+ *    mata file
+ *
+ * (Any failure in above steps is considered as update failed,
+ *  and the file context will be keepped as is before the update
+ *  operation)
+ *
+ * After previous step the update is considered complete, but
+ * we should do the following clean-up step(s):
+ *
+ *  - Unlink all REE files represents the old TEE file (including
+ *    meta and block files)
+ *
+ * (Any failure in above steps is considered as a successfully
+ *  update)
+ */
 int tee_fs_common_rename(const char *old, const char *new)
 {
 	int res = -1;
-	char *tmp = NULL;
-	struct tee_fs_rpc head = { 0 };
-	size_t len_old = strlen(old) + 1;
-	size_t len_new = strlen(new) + 1;
-	size_t len = len_old + len_new;
+	size_t old_len = strlen(old) + 1;
+	size_t new_len = strlen(old) + 1;
+	struct tee_fs_dir *old_dir;
+	struct tee_fs_dirent *dirent;
+	char *meta_filename = NULL;
 
-	if (len > TEE_FS_NAME_MAX)
+	DMSG("old=%s, new=%s", old, new);
+
+	if (old_len > TEE_FS_NAME_MAX || new_len >TEE_FS_NAME_MAX)
 		goto exit;
 
-	tmp = malloc(len);
-	if (!tmp)
+	res = ree_fs_mkdir(new,
+			TEE_FS_S_IRUSR | TEE_FS_S_IWUSR);
+	if (res)
 		goto exit;
-	memcpy(tmp, old, len_old);
-	memcpy(tmp + len_old, new, len_new);
 
-	head.op = TEE_FS_RENAME;
+	old_dir = tee_fs_common_opendir(old);
+	if (!old_dir)
+		goto exit;
 
-	res = tee_fs_send_cmd(&head, tmp, len, TEE_FS_MODE_IN);
-	if (!res)
-		res = head.res;
+	dirent = tee_fs_common_readdir(old_dir);
+	while (dirent) {
+		if (!strncmp(dirent->d_name, "meta.", 5)) {
+			meta_filename = strdup(dirent->d_name);
+		} else {
+			res = create_hard_link(old, new, dirent->d_name);
+			if (res)
+				goto exit_close_old_dir;
+		}
+
+		dirent = tee_fs_common_readdir(old_dir);
+	}
+
+	/* finally, link the meta file, rename operation completed */
+	TEE_ASSERT(meta_filename);
+	res = create_hard_link(old, new, meta_filename);
+	if (res)
+		goto exit_close_old_dir;
 
+	/* we are safe now, remove old TEE file */
+	unlink_tee_file(old);
+
+exit_close_old_dir:
+	tee_fs_common_closedir(old_dir);
 exit:
-	free(tmp);
+	free(meta_filename);
 	return res;
 }
 
+/*
+ * To ensure atomic unlink operation, we can simply
+ * split the unlink operation into:
+ *
+ *  - rename("file", "file.trash");
+ *
+ * (Any failure in above steps is considered as update failed,
+ *  and the file context will be keepped as is before the update
+ *  operation)
+ *
+ * After previous step the update is considered complete, but
+ * we should do the following clean-up step(s):
+ *
+ *  - unlink("file.trash");
+ *
+ * (Any failure in above steps is considered as a successfully
+ *  update)
+ */
 int tee_fs_common_unlink(const char *file)
 {
 	int res = -1;
-	struct tee_fs_rpc head = { 0 };
-	size_t len = strlen(file) + 1;
+	char trash_file[TEE_FS_NAME_MAX + 6];
 
-	if (len > TEE_FS_NAME_MAX)
-		goto exit;
+	snprintf(trash_file, REE_FS_NAME_MAX + 6, "%s.trash",
```

Comment:
Here is a potential overflow buffer problem. I'd guess you intended to write `TEE_FS_NAME_MAX` here also? And not `REE_...`?